### ISSUE TITLE
Long.MIN_VALUE considered out of range for type `long` in range aggregations

### ISSUE THREAD

**Issue by yrodiere** [elastic/elasticsearch#81529]
**Elasticsearch version** (`bin/elasticsearch --version`): 7.16.0

**Plugins installed**: []

**JVM version** (`java -version`):

**OS version** (`uname -a` if on a Unix-like system):

**Description of the problem including expected versus actual behavior**:

When defining a range aggregation on a field of type `long`, if one of the ranges has a bound equal to `Long.MIN_VALUE` (the java constant, equal to `-(2^63)`) or just a bit higher, Elasticsearch raises an error:

```
"Value [-9.223372036854776E18] is out of range for a long"
```

This directly contradicts the [documentation of numeric types](https://www.elastic.co/guide/en/elasticsearch/reference/current/number.html) which states that a `long` is "A signed 64-bit integer with a minimum value of -2^63 and a maximum value of 2^63-1"

It is a regression compared to 7.15.2, which didn't raise such an error.

**Steps to reproduce**:

```
curl -XDELETE -H "Content-Type: application/json" 'localhost:9200/test/' || true
echo
curl -XPUT -H "Content-Type: application/json" 'localhost:9200/test/' -d'{"mappings": {"properties": {"mylong": {"type": "long" } } } }'
echo
curl -XPUT -H "Content-Type: application/json" 'localhost:9200/test/_doc/1?refresh=true' -d '{"mylong": -9223372036854775808 }'
echo
curl -XPOST -H "Content-Type: application/json" 'localhost:9200/test/_search?size=0&pretty' -d '{"query":{"match_all":{}},"aggregations":{"myagg":{"range":{"field":"mylong","ranges":[{"from":-9223372036854775808,"to":9223372036854775807}]}}}}'
```

Output of the last command with Elasticsearch 7.16.0:

```json{"total":2,"successful":1,"failed":0},"_seq_no":0,"_primary_term":1}
{
  "error" : {
    "root_cause" : [
      {
        "type" : "illegal_argument_exception",
        "reason" : "Value [-9.223372036854776E18] is out of range for a long"
      }
    ],
    "type" : "search_phase_execution_exception",
    "reason" : "all shards failed",
    "phase" : "query",
    "grouped" : true,
    "failed_shards" : [
      {
        "shard" : 0,
        "index" : "test",
        "node" : "MgMAsXzNTxeaMkLvCsZjuw",
        "reason" : {
          "type" : "illegal_argument_exception",
          "reason" : "Value [-9.223372036854776E18] is out of range for a long"
        }
      }
    ],
    "caused_by" : {
      "type" : "illegal_argument_exception",
      "reason" : "Value [-9.223372036854776E18] is out of range for a long",
      "caused_by" : {
        "type" : "illegal_argument_exception",
        "reason" : "Value [-9.223372036854776E18] is out of range for a long"
      }
    }
  },
  "status" : 400
}
```

Output of the last command with Elasticsearch 7.15.2:

```json
{
  "took" : 49,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 1,
      "relation" : "eq"
    },
    "max_score" : null,
    "hits" : [ ]
  },
  "aggregations" : {
    "myagg" : {
      "buckets" : [
        {
          "key" : "-9.223372036854776E18-9.223372036854776E18",
          "from" : -9.223372036854776E18,
          "to" : 9.223372036854776E18,
          "doc_count" : 1
        }
      ]
    }
  }
}
```

**Provide logs (if relevant)**: N/A

**Workaround**:

Sending long values as JSON strings instead of JSON numbers removes the problem completely, apparently bypassing the (invalid) check.

So I suppose this is a deserialization bug, where extreme integer values are rounded and converted to doubles during deserialization.

**Comment by jtibshirani** [comment]
Thank you @yrodiere for reporting this. I was also able to reproduce it in 7.16, but not 7.15. I've looped in our aggregations experts to take a look.

**Comment by not-napoleon** [comment]
I wrote up a test for this and ran it through `git bisect`.  Looks like I broke it in https://github.com/elastic/elasticsearch/pull/78344.  Sorry about that.  I'll start looking into a fix, thanks for reporting it.  I am going to be out of office for a bit for the holidays, so I wouldn't expect much movement on this until after that, just to set expectations.

**Comment by not-napoleon** [comment]
As a work around, you should be able to specify `value_type: double` as part of the range aggregation to kick it into double mode, as opposed to long mode.  I think that's functionally the same as the old behavior.

**Comment by yrodiere** [comment]
>  I'll start looking into a fix

Great, thank you.

> ~As a work around, you should be able to specify `value_type: double` as part of the range aggregation to kick it into double mode, as opposed to long mode. I think that's functionally the same as the old behavior.~
> 
> Actually, I think I spoke too soon, and this doesn't work. Sorry about that.

No problem. As I explained in the issue description, there already is a workaround: encode longs as strings (i.e. `"-9223372036854775808"` instead of `-9223372036854775808`). I'm using that workaround for now.

**Comment by salvatore-campagna** [comment]
It looks like the issue here is caused by this line:

```String stringValue = (value instanceof BytesRef) ? ((BytesRef) value).utf8ToString() : value.toString();```

to be more specific ```value.toString()``` produces `-9.223372036854776E18` which causes the issue...I think because of the "E" into the string representation.

**Comment by salvatore-campagna** [comment]
Closing this after merging #83213 which fixes the issue in version 8.1.0.
